#836 fix - Bring back test to check add Add-On to 'My Favorites'

stupchiy opened the pull request:
Bring back test to check add Add-On to 'My Favorites'

stupchiy committed:
#836: Bring back test to check add extension to 'My Favorites'

stupchiy committed:
Merge branch 'master' into issue_836

stephendonner commented:
Running this locally, it does the right thing for me, but there appears to be (like with issue 844?) a problem with updating data/state, such that even manually refreshing the page doesn't show the add-on was added to that collection (says "0 add-ons in this collection, which the test failure correctly asserts too, as follows):

`(addon-tests) sdonners-MacBook-Pro:Addon-Tests sdonner$ py.test --driver=Firefox --base-url https://addons-dev.allizom.org --servicesbaseurl https://services.addons-dev.allizom.org --variables=/Users/sdonner/Desktop/AMO.json tests/desktop/test_collections.py 
============================= test session starts ==============================
platform darwin -- Python 2.7.11, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /Users/sdonner/Addon-Tests, inifile: setup.cfg
plugins: html-1.8.0, rerunfailures-0.5, selenium-1.2.1, variables-1.4, xdist-1.14
collected 4 items 

tests/desktop/test_collections.py .fxapom deleted the account for email: gkchmamxplxc@restmail.net at https://stable.dev.lcip.org/auth/ on 2016-04-18 23:28:23.455068
.fxapom deleted the account for email: fjqscbnkhjhq@restmail.net at https://stable.dev.lcip.org/auth/ on 2016-04-18 23:28:35.832916
.F

=================================== FAILURES ===================================
_________________ TestCollections.test_user_my_favorites_page __________________

self = <tests.desktop.test_collections.TestCollections instance at 0x107e9c0e0>
base_url = 'https://addons-dev.allizom.org'
selenium = <selenium.webdriver.firefox.webdriver.WebDriver (session="60e5edb2-b572-be46-864d-a73763e24b3a")>
logged_in = None

    @pytest.mark.native
    def test_user_my_favorites_page(self, base_url, selenium, logged_in):
        home_page = Home(base_url, selenium)
        my_collections_page = home_page.header.click_my_collections()
    
        my_favorites_page = my_collections_page.click_my_favorites()
        assert my_favorites_page.favorites_add_ons_count == 0
    
        assert 'My Favorite Add-ons' == my_favorites_page.my_favorites_header_text
        edit_collection_page = my_favorites_page.edit_collection()
    
        assert edit_collection_page.is_the_current_page
    
        edit_collection_page.click_add_ons_tab()
        edit_collection_page.include_add_on('Firebug')
        edit_collection_page.click_add_ons_save_changes()
    
        my_favorites_page = home_page.header.click_my_favorites()
>       assert my_favorites_page.favorites_add_ons_count > 0
E       assert 0 > 0
E        +  where 0 = <pages.desktop.user.MyFavorites object at 0x10198d810>.favorites_add_ons_count

tests/desktop/test_collections.py:66: AssertionError
---------------------------- Captured stdout setup -----------------------------
fxapom created an account for email: gqukfndjsfql@restmail.net at https://stable.dev.lcip.org/auth/ on 2016-04-18 23:28:38.805703
------------------------------- pytest-selenium --------------------------------
URL: https://addons-dev.allizom.org/en-US/firefox/collections/gqukfndjsfql/favorites/
===================== 1 failed, 3 passed in 71.12 seconds ======================`

davehunt commented:
Can you see if you have the same issues when running on https://addons.allizom.org

stupchiy commented:
oh, it is passing there @davehunt

stephendonner commented:
Yep, can confirm - this pull/test works just fine (consistently) against https://addons.allizom.org, but fails against https://addons-dev.allizom.org

davehunt commented:
@stupchiy can you add an xfail when this is run on dev?

stupchiy committed:
Merge branch 'master' into issue_836

stupchiy commented:
oh, sure @davehunt is there a way to do that? cool, can you point me to example, haven't found smth like that in Addon-Tests

davehunt commented:
Here's an example: https://github.com/mozilla/Addon-Tests/blob/53d65a2538eb9830f7b17ebb721ab732da25ac71/tests/desktop/test_reviews.py#L49-L51

stupchiy committed:
marked test_user_my_favorites_page as @xfail for addons-dev environment

stupchiy committed:
Merge branch 'master' into issue_836

stupchiy commented:
@davehunt updated, please check if make sense. thx

davehunt commented:
@stupchiy the xfail looks fine, but I haven't looked through the other changes in the patch. If someone else has some time to review it that would be great, otherwise I'll try to get to it tomorrow.

stupchiy commented:
cool thx @davehunt , @stephendonner, @m8ttyB r?

stephendonner commented:
Sorry, didn't get time for this (blog post + https://github.com/mozilla/mozillians-tests/pull/237 + Sauce failures).  Punting over to Dave :-\

davehunt commented on tests/desktop/test_collections.py:
```diff
@@ -44,3 +44,28 @@ def test_user_my_collections_page(self, base_url, selenium, logged_in, user):
+    @pytest.mark.native
+    @pytest.mark.xfail(
+        'addons-dev' in pytest.config.getvalue('base_url'),
+        reason='Problem with updating state on addons-dev')
```
We should reference the issue in addons-server, or perhaps raise a new issue as this is similar to https://github.com/mozilla/addons-server/issues/2442 but may be a different bug.

davehunt commented on tests/desktop/test_collections.py:
```diff
@@ -44,3 +44,28 @@ def test_user_my_collections_page(self, base_url, selenium, logged_in, user):
+        edit_collection_page.click_add_ons_save_changes()
+
+        my_favorites_page = home_page.header.click_my_favorites()
+        assert my_favorites_page.favorites_add_ons_count > 0
```
We should be able to assert this is exactly 1

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +165,106 @@ def clear_field(self):
+    _add_on_name_locator = (By.CSS_SELECTOR, '.item h3 a')
+    _add_to_favorites_locator = (By.CSS_SELECTOR, 'a.favorite')
+
+    def __init__(self, base_url, selenium, user_name=None):
```
I don't think we need to overload the `__init__`. As far as I can tell we're not using the `user_name` kwarg anywhere..

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +165,106 @@ def clear_field(self):
+        self.selenium.find_element(*tab_locator).click()
+
+    def click_add_ons_tab(self):
+        tab_locator = (By.CSS_SELECTOR, ".tab-nav li:nth-child(2) a")
```
We should keep all locators together. Please rename this to `_addons_tab_locator` and move it to be with the others.

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +165,106 @@ def clear_field(self):
+        WebDriverWait(self.selenium, self.timeout).until(
+            lambda s: self.is_element_visible(*self._title_locator))
+
+    def click_description_tab(self):
```
This is not used, we should avoid the temptation of adding methods until we have code that uses them.

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +165,106 @@ def clear_field(self):
+        self.selenium.find_element(self._save_changes_locator[0],
+                                   self._save_changes_locator[1] % locator).click()
+
+    def click_description_save_changes(self):
```
Remove this unused function.

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +165,106 @@ def clear_field(self):
+    def click_add_ons_save_changes(self):
+        locator = '#addons-edit'
+        self.selenium.find_element(self._save_changes_locator[0],
+                                   self._save_changes_locator[1] % locator).click()
```
Create a locator named `_save_addon_changes_locator` and avoid having to format a locator inline.

davehunt commented:
r+ with comments addressed (though I haven't run the new test locally)

This is not necessary enough to block, as the patch currently matches the conventions used in the suite, but I would like to see more use of page regions. This allows the tests to be more readable, and the page objects more flexible. For example, the same test could be written as:

```python
def test_add_addon_to_my_favorites(self, base_url, selenium, logged_in):
    page = Home(base_url, selenium)
    my_collections_page = home_page.header.click_my_collections()
    my_favorites_page = my_collections_page.click_my_favorites()
    assert 0 == len(my_favorites_page.add_ons)

    edit_collection_page = my_favorites_page.edit_collection()
    edit_collection_page.click_add_ons_tab()
    edit_collection_page.include_add_on('Firebug')
    edit_collection_page.click_add_ons_save_changes()

    my_favorites_page = home_page.header.click_my_favorites()
    assert 1 == len(my_favorites_page.add_ons) > 0
    assert 'Firebug' == my_favorites_page.add_ons[0].name
    assert my_favorites_page.add_ons[0].favorite
```

This would require an `AddOn` region for each of the add-ons listed in the collection, but gives the flexibility of multiple add-ons being in the collection, and for the tests to address any of them.

stupchiy committed:
updated according to comments from @davehunt, added AddOn region usage for collections

stupchiy commented:
hey @davehunt, I corrected mentioned by you mistakes, also added a new issue https://github.com/mozilla/addons-server/issues/2629. I have implemented usage of `AddOn` region, but not sure it is ok, since I needed to extend it a bit, due to different title locator, would be happy to get a feedback.

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +166,78 @@ def clear_field(self):
+    def title(self):
+        return self.selenium.find_element(*self._title_locator).text
+
+    def __init__(self, base_url, selenium, collection_name=None, user_name=None):
```
I still don't see where collection_name and user_name are used? I don't think we need these at all.

davehunt commented on tests/desktop/test_collections.py:
```diff
@@ -44,3 +44,27 @@ def test_user_my_collections_page(self, base_url, selenium, logged_in, user):
+        edit_collection_page.click_add_ons_save_changes()
+
+        my_favorites_page = home_page.header.click_my_favorites()
+        assert 1 == len(my_favorites_page.add_ons) > 0
```
Remove the trailing `> 0`

davehunt commented on pages/desktop/user.py:
```diff
@@ -165,8 +166,78 @@ def clear_field(self):
+    @property
+    def add_ons(self):
+        return [AddOn(self.base_url, self.selenium, el) for el in
+                self.selenium.find_elements(*self._addon_locator)]
```
I wasn't expecting you to implement this, but thanks for giving it a go! We can't actually reuse `AddOn` from `pages.desktop.regions.addon` because that's a different region. Here I would suggest a nested class named `AddOn` as the region is specific to the collection page.

davehunt commented:
Thanks @stupchiy! I really wasn't expecting the AddOn region changes so it's awesome that you've included those. Unfortunately the existing region isn't suitable here, so we would be better of creating a new region specifically for use in this page class.

stupchiy committed:
removed xfail(passing on addons-dev now), removed unused functions, added nested AddOn PageRegion

stupchiy commented:
no worries @davehunt , always open for challenges ;-) I've removed xfail, can confirm is passing now. Also made adoptions to PageRegion...

stupchiy committed:
Merge branch 'master' into issue_836

davehunt commented:
Thanks @stupchiy!
